sdks/python/apache_beam/runners/common.py

```diff
@@ -765,41 +978,63 @@ def _invoke_process_per_window(self,
             element, deferred_restriction)
-        return ((
-            windowed_value.with_value(((element, deferred_restriction), size)),
-            output_watermark), deferred_watermark)
-    return None
+        if self.threadsafe_watermark_estimator:
+          output_watermark = (
+              self.threadsafe_watermark_estimator.current_watermark())
+          estimator_state = (
+              self.threadsafe_watermark_estimator.get_estimator_state())
+          return ((
+              windowed_value.with_value(
+                  ((element, (deferred_restriction, estimator_state)), size)),
+              output_watermark), deferred_watermark)
+        else:
+          return ((
+              windowed_value.with_value(
+                  ((element, deferred_restriction), size)), None),
+                  deferred_watermark)
+        return None
 
   def try_split(self, fraction):
     # type: (...) -> Optional[Tuple[SplitResultType, SplitResultType]]
-    if self.threadsafe_restriction_tracker and self.current_windowed_value:
-      # Temporary workaround for [BEAM-7473]: get current_watermark before
-      # split, in case watermark gets advanced before getting split results.
-      # In worst case, current_watermark is always stale, which is ok.
-      if self.watermark_estimator:
-        current_watermark = self.watermark_estimator.current_watermark()
-      else:
-        current_watermark = None
-      split = self.threadsafe_restriction_tracker.try_split(fraction)
+    restriction_tracker = self.threadsafe_restriction_tracker
+    current_windowed_value = self.current_windowed_value
+    if restriction_tracker and current_windowed_value:
+      current_watermark = None
+      # Make sure that the RestrictionTracker and WatermarkEstimator are locked
+      # together.
+      with self._synchronized_lock:
+        split = restriction_tracker.try_split(fraction)
+        if self.threadsafe_watermark_estimator:
+          current_watermark = (
+              self.threadsafe_watermark_estimator.current_watermark_with_lock())
+          estimator_state = (self.threadsafe_watermark_estimator
+                             .get_estimator_state_with_lock())
       if split:
         primary, residual = split
         element = self.current_windowed_value.value
         restriction_provider = self.signature.get_restriction_provider()
         primary_size = restriction_provider.restriction_size(element, primary)
         residual_size = restriction_provider.restriction_size(element, residual)
-        return (
-            ((self.current_windowed_value.with_value((
-                (element, primary), primary_size)), None), None),
-            ((self.current_windowed_value.with_value((
-                (element, residual), residual_size)), current_watermark), None))
+        if self.threadsafe_watermark_estimator:
+          return (
+              ((self.current_windowed_value.with_value((
```

Comment:
As mentioned in the typing PRs, these tuples of tuples of tuples are getting a bit unwieldy. Perhaps we could introduce a class (or at least named tuple)? Open question: should the primary the same type from the residual, or should it be a different type (given it can hold a watermark and delay time).

Reply:
I like the idea of named tuple. I also agree the primary should be different from the residual. If that sounds good, I can make these changes together with moving `ThreadsafeRestrictionTracker` out from `common.py`.